File: java/tools/src/main/java/org/artrev/workshop/fp/tools/flow/ZippingFlow.java
package org.artrev.workshop.fp.tools.flow;

import org.artrev.workshop.fp.tools.Function2;
import org.artrev.workshop.fp.tools.control.Try;

import java.util.NoSuchElementException;

class ZippingFlow<A, B, C> implements Flow<C> {
    private final Flow<A> left;
    private final Flow<B> right;
    private final Function2<A, B, C> zipper;

    ZippingFlow(final Flow<A> left,
                final Flow<B> right,
                final Function2<A, B, C> zipper) {
        this.left = left;
        this.right = right;
        this.zipper = zipper;
    }

    @Override
    public Try<C> next() {
        final Try<A> nextLeft = left.next();
        if (nextLeft.isFailure())
            return Try.failure(new NoSuchElementException("Left flow is out of elements to zip!"));

        final Try<B> nextRight = right.next();
        if (nextRight.isFailure())
            return Try.failure(new NoSuchElementException("Right flow is out of elements to zip!"));

        return nextLeft.flatMap(a -> nextRight.flatMap(b -> Try.success(zipper.apply(a, b))));
    }
}

File: kotlin/tools/src/main/kotlin/org/artrev/workshop/fp/tools/flow/ZippingFlow.kt
@file:Suppress("unused")

package org.artrev.workshop.fp.tools.flow

import io.github.ajoz.workshop.fp.tools.control.Try
import java.util.NoSuchElementException

/**
 *
 */
fun <A, B, C> Flow<A>.zip(other: Flow<B>, function: (A, B) -> C): Flow<C> {
    class ZippingFlow<A, B, C>(private val left: Flow<A>,
                               private val right: Flow<B>,
                               private val zipper: (A, B) -> C) : Flow<C> {

        override fun next(): Try<C> {
            val nextLeft = left.next()
            if (nextLeft.isFailure)
                return Try.Failure(NoSuchElementException("Left flow is out of elements to zip!"))

            val nextRight = right.next()
            return if (nextRight.isFailure)
                Try.Failure(NoSuchElementException("Right flow is out of elements to zip!"))
            else
                nextLeft.flatMap { a ->
                    nextRight.flatMap { b ->
                        Try.Success(zipper(a, b))
                    }
                }

        }
    }

    return ZippingFlow(this, other, function)
}
